### PR TITLE
[JBPM-9654] Migration of all subprocess of a parent process fails

### PR DESCRIPTION
**Pull request by elguardian**
Jira : https://issues.redhat.com/browse/JBPM-9654

### CODE REVIEW

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java [hunk 1]
```diff
@@ -101,21 +102,40 @@ public MigrationReportInstanceList migrateProcessInstances(String containerId, L
...
+            ProcessInstanceDesc piToMigrate = runtimeDataService.getProcessInstanceById(processInstanceToMigrateId);
+            Optional<MigrationProcessSpecification> spec = migrationSpecification.getProcesses().stream().filter(e -> piToMigrate.getProcessId().equals(e.getSourceProcessId())).findFirst();
+            if(!spec.isPresent()) {
+                logger.error("MigrationProcessSpecification is not correct. Process Instance Id " + processInstanceToMigrateId + " won't be migrated");
+                continue;
+            }
+            String targetProcessId = spec.get().getTargetProcessId();
+            Map<String, String> nodeMapping = spec.get().getNodes();
+            reports.add(processInstanceMigrationService.migrate(containerId, processInstanceToMigrateId, targetContainerId, targetProcessId, nodeMapping));
+        }
```
Comment (afalhambra) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java:136]: Are we not missing the parent instance migration here? 
I mean, even if we pass parentProcessInstance as an `MigrationProcessSpecification` object in the request, we are not taking into account the parentInstanceId to be migrated.
Below we are only looping through the subprocess instances as result of the query `getProcessInstancesWithSubprocessByProcessInstanceId` (which does not return the parent)

I believe we should also migrate the parentInstanceId after the loop.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java [hunk 2]
```diff
@@ -266,4 +284,92 @@ public void testUpgradeProcessInstancesWithNodeMapping() throws Exception {
...
+            assertEquals(2, tasks.size());
+            for (TaskSummary task : tasks) {
+                assertEquals("Evaluate items?", task.getName());
+                assertEquals(CONTAINER_ID, task.getContainerId());
+                assertEquals(SUBPROCESS_ID, task.getProcessId());
+            }
+
+            MigrationSpecification spec = new MigrationSpecification();
+            MigrationProcessSpecification pSpecParent = new MigrationProcessSpecification();
+            pSpecParent.setSourceProcessId(PARENT_WITH_INDEPENDENT_SUBPROCESS_ID);
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java:317]: Source process Id is `PARENT_WITH_NON_INDEPENDENT_SUBPROCESS_ID`
```suggestion
            pSpecParent.setSourceProcessId(PARENT_WITH_NON_INDEPENDENT_SUBPROCESS_ID);
```

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java [hunk 3]
```diff
@@ -101,21 +102,40 @@ public MigrationReportInstanceList migrateProcessInstances(String containerId, L
...
+
```
Comment (afalhambra) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java:137]: Shouldn't we also check for the parent presence, in the `MigrationSpecification` object and throw an exception if it's not coming?

**Review by afalhambra — CHANGES_REQUESTED** [review]
Just one comment about parent instance to be migrated. Rest are minor issues, will add some tests on top if you don't mind.
Thanks!

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/admin/MigrationProcessSpecification.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        return targetProcessId;
+    }
+
+    public void setTargetProcessId(String targetProcessId) {
+        this.targetProcessId = targetProcessId;
+    }
+
+    public Map<String, String> getNodes() {
+        if(nodes == null) {
+            emptyMap();
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/admin/MigrationProcessSpecification.java:56]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java [hunk 5]
```diff
@@ -266,4 +284,92 @@ public void testUpgradeProcessInstancesWithNodeMapping() throws Exception {
...
+        Long processParentInstanceId = processClient.startProcess(CONTAINER_ID, PARENT_WITH_NON_INDEPENDENT_SUBPROCESS_ID);
+        assertNotNull(processParentInstanceId);
+        assertTrue(processParentInstanceId > 0);
+
+        try {
+            logger.info("Process in container {} has started {}", CONTAINER_ID, processParentInstanceId);
+            List<ProcessInstance> childrenProcessInstances = processClient.findProcessInstancesByParent(CONTAINER_ID, processParentInstanceId, 0, 10);
+            assertEquals(2, childrenProcessInstances.size());
+            for (ProcessInstance processInstance : childrenProcessInstances) {
+                logger.info("Subrocess in container {} has started {}", CONTAINER_ID, processInstance.getId());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java:300]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java [hunk 6]
```diff
@@ -266,4 +284,92 @@ public void testUpgradeProcessInstancesWithNodeMapping() throws Exception {
...
+            MigrationProcessSpecification pSpecChild = new MigrationProcessSpecification();
+            pSpecChild.setSourceProcessId(SUBPROCESS_ID);
+            pSpecChild.setTargetProcessId(SUBPROCESS_ID_2);
+            spec.setProcesses(Arrays.asList(pSpecParent, pSpecChild));
+
+            ProcessInstance piMigrate = processClient.getProcessInstance(CONTAINER_ID, processParentInstanceId);
+            logger.info("about to process migration from container {} to {} with process definintion {} with id {}", piMigrate.getContainerId(), CONTAINER_ID_2, piMigrate.getProcessId(), processParentInstanceId);
+            List<MigrationReportInstance> reports = processAdminClient.migrateProcessInstanceWithSubprocess(CONTAINER_ID, processParentInstanceId, CONTAINER_ID_2, spec);
+            assertNotNull(reports);
+            for(MigrationReportInstance report : reports) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java:328]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java (hunk 2 above)
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java:317]: nope

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java (hunk 3 above)
Comment (elguardian) [reply kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java:137]: nope... we migrate what we can.

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java (hunk 1 above)
Comment (elguardian) [reply kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/admin/ProcessAdminServiceBase.java]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java [hunk 7]
```diff
@@ -266,4 +284,92 @@ public void testUpgradeProcessInstancesWithNodeMapping() throws Exception {
...
+            MigrationProcessSpecification pSpecParent = new MigrationProcessSpecification();
+            pSpecParent.setSourceProcessId(PARENT_WITH_INDEPENDENT_SUBPROCESS_ID);
+            pSpecParent.setTargetProcessId(PARENT_WITH_INDEPENDENT_SUBPROCESS_ID);
+            MigrationProcessSpecification pSpecChild = new MigrationProcessSpecification();
+            pSpecChild.setSourceProcessId(SUBPROCESS_ID);
+            pSpecChild.setTargetProcessId(SUBPROCESS_ID_2);
+            spec.setProcesses(Arrays.asList(pSpecParent, pSpecChild));
+
+            ProcessInstance piMigrate = processClient.getProcessInstance(CONTAINER_ID, processParentInstanceId);
+            logger.info("about to process migration from container {} to {} with process definintion {} with id {}", piMigrate.getContainerId(), CONTAINER_ID_2, piMigrate.getProcessId(), processParentInstanceId);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/admin/ProcessInstanceMigrationIntegrationTest.java:325]: done

**Pull request merged**